jupyter-extension/jupyterlab_pachyderm/pps_client.py

```diff
@@ -41,72 +39,142 @@ async def create(self, path: str, config: dict):
+            raise HTTPError(reason=f"Bad Request: field pipeline_name not set")
+
+        image = config.get('image', None)
+        if image is None:
+            raise HTTPError(reason=f"Bad Request: field image not set")
+
+        requirements = config.get('requirements')
+
+        input_spec = config.get('input_spec', None)
+        if input_spec is None:
+            raise HTTPError(reason=f"Bad Request: field input_spec not set")
+
+        return cls(
+            pipeline_name=pipeline_name,
+            image=image,
+            requirements=requirements,
+            input_spec=input_spec
+        )
+
+
+def create_pipeline_spec(config: PpsConfig, companion_branch: str) -> dict:
+    companion_repo = f"{config.pipeline_name}__context"
+    input_spec = dict(
+        cross=[
+            config.input_spec,
+            {'pfs': dict(repo=companion_repo, branch=companion_branch, glob="/")}
+        ]
     )
-    cmd = ["python3", "-c", micro_entrypoint, script_name]
     return dict(
-        pipeline=dict(name=pipeline_name),
+        pipeline=dict(name=config.pipeline_name),
         description="Auto-generated from notebook",
-        transform=dict(cmd=cmd, image=image),
+        transform=dict(
+            cmd=["python3", f"/pfs/{companion_repo}/entrypoint.py"],
+            image=config.image
+        ),
         input=input_spec,
         update=True,
-        reprocess=True
+        reprocess=True,
     )
+
+
+def upload_environment(
+        client: python_pachyderm.Client, repo: str, config: PpsConfig, script: bytes
+) -> str:
+    """Manages the pipeline's "environment" through a companion repo.
+
+    The companion repo contains:
+      - user_code.py     : the user's python script
+      - requirements.txt : requirements optionally specified by the user
+      - entrypoint.py    : the entrypoint for setting up and running the user code
+      and the entrypoint script to orchestrate their code.
+
+    Returns
+    -------
+        The branch name with the new commit as HEAD. We need to create a new branch
+          as a workaround because specifying a commit within a pipeline's input
+          spec is currently unsupported.
+    """
+
+    def entrypoint():
+        """Entrypoint used by the PPS extension."""
+        print("Greetings from the Pachyderm PPS Extension")
+
+        from pathlib import Path
+        from subprocess import run
+
+        reqs = Path("requirements.txt")
+        if reqs.exists():
+            run(["pip", "--disable-pip-version-check", "install", "-r", reqs.as_posix()])
+
+        import user_code  # This runs the user's code.
+
+    entrypoint_script = (
+        f'{dedent(getsource(entrypoint))}\n\n'
+        'if __name__ == "__main__":\n'
+        '    entrypoint()\n'
+    )
+
+    with client.commit(repo, "master") as commit:
+        # Remove the old files.
+        for item in client.list_file((repo, "master"), "/"):
+            client.delete_file(commit, item.file.path)
```

Comment:
Why not make a new commit on top?

Reply:
Not sure what you mean by "commit on top". This _is_ a little heavy handed, but it's to make sure that no files linger in the context when the pipeline is updated. A example which could cause trouble is a user removing their "requirements.txt" file and we want to make sure its deleted in the companion repo. This will also be more important when we allow users to include arbitrary files in their "context"

Reply:
This makes sense, sorry my mental model was wrong here (was thinking of separate branches, not commits, which are of course additive)